### DISCUSSION TITLE
More control over which editors are closed when using "Open With"

### DISCUSSION THREAD

**Discussion by ptziegler** [eclipse-platform/eclipse.platform.ui#2191] [Ideas]
In our application, we have limited the amount of editors that can be opened at the same time to twelve. Once this limit is reached, the user is given a list of all open editors and asked to select the ones that should be closed.

![image](https://github.com/user-attachments/assets/8bfb513f-880b-47f4-ad0a-2501d1e96955)

Sadly, this doesn't work when using "Open With" after right-clicking on a file, as the underlying implementation is internal. The editor-to-close is determined via the [TabBehavior](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/tweaklets/TabBehaviour.java), more specifically the [TabBehaviorMRU](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/tweaklets/TabBehaviourMRU.java). Here the least recently editor is selected (with special handling of pinned/dirty editors), without the chance of any user interaction.

This behavior is leads to inconsistencies in our project and to solve it, I'd like to have more control over which editors are selected, rather than simply picking the oldest one. After thinking about this for a bit, I came up with two possible solutions, with my personal preference being the second one:

- Create a new TabBehavior class for this use-case and have it be configurable via e.g. the preference page
- Make (parts) of the TabBehavior component public, so that users can contribute their own implementation

One could simply define an `ITabBehaviour` interface, which declares the `findReusableEditor(IWorkbenchPage page)` methods. Implementation of this class could either be contributed via an extension point or a plain OSGi service. Alternatively, one also define a static `setDefault(ITabBehaviour)` method (similar to what already exists for the [Tweaklets](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/tweaklets/Tweaklets.java)). If no other contribution exists, one would then default to the `TabBehaviourMRU` for the current behavior.

Alternatively, I could also simply ignore the "internal" warning and implement the TabBehavior directly. Though I'd like to avoid that unless it's unavoidable...

Do you have any concerns regarding this approach or perhaps know of a better way to do this?

**Comment by ptziegler** [comment]
Hm... the `TabBehaviourMRU` is calling internal methods such as `getSortedEditors()`, `getEditorReuseThreshold()` and `closeEditor(IEditorReference, boolean)`. So it's necessary to work on the internal WorkbenchPage object, rather than the IWorkbenchPage interface. ☹️

**Reply by laeubi** [reply]
It seems all of these mentioned things are already heavily internal API anyways. I wonder if you maybe better can handle the case using the E4 Model where you can get events when new items are created and you can also close parts if desired and implement any behavior you want. Its a bit brittle to getting started, but once done it is quite powerful and flexible.

**Reply by ptziegler** [reply]
> I wonder if you maybe better can handle the case using the E4 Model where you can get events when new items are created and you can also close parts if desired and implement any behavior you want.

It's been a while since I dabbled with the E4 model, but I think something like this could work. If I can inject some custom logic (via event or some other mechanism) _before_ the TabBehaviour code is executed, then that would give me the desired behavior.

I assume that's what you mean? Use e.g. the IEventBroker, subscribe to the desired event and then act upon them?

**Reply by laeubi** [reply]
Yes if you simply allow unlimited editors and implement the limiation yourself (e.g. you can count the number of childs in primary datastack whenever its childcount changes).

As a first guess, maybe `org.eclipse.e4.ui.workbench.UIEvents.ElementContainer.TOPIC_CHILDREN`

**Reply by ptziegler** [reply]
> As a first guess, maybe org.eclipse.e4.ui.workbench.UIEvents.ElementContainer.TOPIC_CHILDREN

That seems to work and after some testing, I came up with following prototype, which defines a limit of 5 editors:

```
public class OpenEditorAddon {
    @Inject
    public void init(IEventBroker eventBroker, EPartService partService) {
        eventBroker.subscribe(UIEvents.ElementContainer.TOPIC_CHILDREN, event -> {
            if (UIEvents.isADD(event) && event.getProperty(UIEvents.EventTags.ELEMENT) instanceof MPartStack partStack) {
                if (partStack.getChildren().size() >= 5) {
                    partService.hidePart((MPart)partStack.getChildren().get(4), true);
                }
            }
        });
    }
}
```

There is obviously still more work that needs to be done, but you've already helped me out quite a bit. Thank you a lot! 😄

**Reply by laeubi** [reply]
Just a nitpick: As a good citizen you should unregister from the broker at some time, or you can register to the event in [a declarative way](https://wiki.eclipse.org/Eclipse4/RCP/Event_Model#Dependency_Injection).
